sentry_sdk/utils.py

```diff
@@ -226,8 +226,21 @@ def to_timestamp(value):
 
 
 def format_timestamp(value):
+    """Formats a timestamp as a RFC3339 string, including timezone information.
+
+    If the timezone is UTC, the string will end with a 'Z' instead of '+00:00'. Otherwise, the timezone offset is
+    included as ±HH:MM.
+
+    If the input is a naive datetime (i.e. without timezone information), the timezone offset is omitted. Note that
+    Relay will interpret such timestamps as UTC, so it is recommended to always use timezone-aware datetimes; otherwise,
+    you may observe unexpected behavior.
+    """
     # type: (datetime) -> str
-    return value.strftime("%Y-%m-%dT%H:%M:%S.%fZ")
+    tz_offset_format = "%:z"  # format specifier for ±HH:MM for non-UTC timezones (or empty string for naive datetimes)
+    if value.tzinfo == timezone.utc:
+        tz_offset_format = "Z"  # literal 'Z' for UTC
+
+    return value.strftime(f"%Y-%m-%dT%H:%M:%S.%f{tz_offset_format}")  # noqa: E231
```

Comment:
Don't need to assemble this manually, [`datetime.isoformat()`](https://docs.python.org/3.11/library/datetime.html#datetime.datetime.isoformat) does this.

Reply:
Yeah, looks like you're right.

I think, however, we should make the code continue to emit the same formatting for UTC times; i.e. with a "Z" at the end rather than "+00:00". Otherwise, on the off chance anything depends on this behavior, it will continue to work. To do this, we will still need the manual check against UTC. And, we will want to pass `"microseconds"` as the `timespec` argument to continue getting microseconds as we have until now.

Also, we need to be careful because ISO 8601 supports timestamp strings that RFC 3339 (which Relay uses) does not. It looks like `datetime.isoformat()` for the most part produces RFC 3339-compatible timestamps, but one difference I noted is that `isoformat()` may include [seconds and microseconds in the UTC offset](https://docs.python.org/3.11/library/datetime.html#datetime.datetime.isoformat) (I am assuming when they are nonzero, which is rare but [possible](https://datatracker.ietf.org/doc/html/rfc3339#section-4.2)). RFC 3339 can [only represent UTC](https://datatracker.ietf.org/doc/html/rfc3339#section-5.6) offsets with an integer amount of hours and minutes, no seconds or microseconds allowed. To get around this, we need to truncate the `isoformat()` output to no more than 32 characters, as this is the exact length of an `isoformat` timestamp without the UTC offset seconds and microseconds fields.